app/services/api_authentication_system.rb

```diff
@@ -19,4 +23,21 @@ def permissions(action_name, resource)
     end
   end
 
+  # Override "login_from_api_key" so that if no key is given or the given key is
+  # invalid, the user is logged in as the guest user.
+  def login_from_api_key
+    key = params[:key]
+    if key.nil?
+      u = User.find_by_login('guestuser')
+    else
+      u = User.find_by_apikey(params[:key]) || User.find_by_login('guestuser')
```

Comment:
If a key is given and it is wrong, I think it should return an error that states it is an invalid login. Otherwise the user might be confused why their login does not allow them to do certain actions in case the API is wrong (since it becomes guestuser).

Reply:
Better?